File: 2_algorithms/grokking-algorithms/2_binary_search/c/binary_search.c
#include <stdio.h>

int* binary_search(int arr[], int size, int item) {

    int low = 0;
    int high = size - 1;

    while (low <= high){
        
        int midd = low + (high - low) / 2;

        if (arr[midd] == item){
            return &arr[midd];
        }

        if (arr[midd] < item){
            low = midd + 1;
        } else {
            high = midd - 1;
        }
    }

    return NULL;
}

int main() {

    int arr[] = {1, 2, 3, 4, 5, 6, 7, 8, 9, 10};
    int size = sizeof(arr) / sizeof(arr[0]);
    int item = 10;

    int* result = binary_search(arr, size, item);
    {
        if (result != NULL) {
            printf("Element %d found at index %d\n", item, result - arr);
        } else {
            printf("Element %d not found\n", item);
        }
    }
}

File: 2_algorithms/grokking-algorithms/2_binary_search/go/binary_search.go
package grokkingalgorithms

import "fmt"

func binarySearch(arr []int, target int) *int {

	var (
		low  = 0
		high = len(arr) - 1
	)

	for low <= high {
		
		mid := (low + high) / 2

		if arr[mid] == target {
			return &mid
		}

		if arr[mid] < target {
			low = mid + 1
		} else {
			high = mid - 1
		}
	}

	return nil
}

func BinarySearch() {

	r := binarySearch([]int{1, 2, 3, 4, 5, 6, 7, 8, 9, 10}, 9)
	{
		if r == nil {
			fmt.Println("Not found")
		} else {
			fmt.Println(*r)
		}
	}
}
